blob/drivertest/drivertest.go

```diff
@@ -414,6 +417,83 @@ func testWrite(t *testing.T, newHarness HarnessMaker, pathToTestdata string) {
 	}
 }
 
+// testCanceledWrite tests the functionality of canceling an in-progress write.
+func testCanceledWrite(t *testing.T, newHarness HarnessMaker, pathToTestdata string) {
+	const key = "blob-for-canceled-write"
+	smallText := []byte("hello world")
+	chunkableContent := loadTestFile(t, pathToTestdata, "test-chunkable.tar")
+
+	tests := []struct {
+		description string
+		content     []byte
+		contentType string
+	}{
+		{
+			// The partial write will be buffered in the concrete type as part of
+			// ContentType detection, so the first call to the Driver will be Close.
+			description: "EmptyContentType",
+			content:     smallText,
+		},
+		{
+			// The partial write will be sent to the Driver, which may do its own
+			// internal buffering.
+			description: "NonEmptyContentType",
+			contentType: "text/plain",
+			content:     smallText,
+		},
+		{
+			// A write larger than the BufferSize we set below, so some providers
+			// make start uploading the first chunk before the cancel.
+			description: "ChunkedUpload",
+			content:     chunkableContent,
+		},
+	}
+
+	ctx := context.Background()
+	for _, test := range tests {
+		t.Run(test.description, func(t *testing.T) {
+			cancelCtx, cancel := context.WithCancel(ctx)
+			h, err := newHarness(ctx, t)
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer h.Close()
+			b, err := h.MakeBucket(ctx)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// Create a writer with the context that we're going
+			// to cancel.
+			opts := &blob.WriterOptions{
+				ContentType: test.contentType,
+				BufferSize:  8 * 1024 * 1024,
```

Comment:
Add a comment saying this should be the minimum allowed buffer size and varies between providers. So that when another provider is added, they can adjust it to have the expected behavior.

Reply:
Done.

Reply:
(in PR #476 )